docs/config.rst

```diff
@@ -317,15 +317,50 @@ your documents.
             </doctrine_mongodb:config>
         </container>
 
-    .. code-block:: php
+You can also use a service as custom type by providing the service id. It's
+necessary to enable the ``type_registry`` by setting.
 
-        use Symfony\Config\DoctrineMongodbConfig;
+.. configuration-block::
 
-        return static function (DoctrineMongodbConfig $config): void {
-            $config->type('custom_type')
-                ->class(\Fully\Qualified\Class\Name::class)
-            ;
-        }
+    .. code-block:: yaml
+
+        services:
+            app.custom_type_service:
+                class: Fully\Qualified\Class\Name
+
+        doctrine_mongodb:
+            type_registry: true
+            types:
+                custom_type: @app.custom_type_service
+
+    .. code-block:: xml
+
+        <?xml version="1.0" ?>
+
+        <container xmlns="http://symfony.com/schema/dic/services"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xmlns:doctrine_mongodb="http://symfony.com/schema/dic/doctrine/odm/mongodb"
+                   xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd
+                                        http://symfony.com/schema/dic/doctrine/odm/mongodb https://symfony.com/schema/dic/doctrine/odm/mongodb/mongodb-1.0.xsd">
+
+            <services>
+                <service id="app.custom_type_service" class="Fully\Qualified\Class\Name" />
+            </services>
+
+            <doctrine_mongodb:config
+                type-registry="true"
+            >
+                <doctrine_mongodb:type name="custom_type" service="app.custom_type_service" />
+            </doctrine_mongodb:config>
+        </container>
+
+.. note::
+
+    Defining a custom type as a service requires Doctrine MongoDB ODM 2.16 or higher.
```

Comment:
Is it "custom type as a service" or "service as custom type"? I think you should pick one and stick to it, and "custom type as a service" sounds more correct.